src/runtime/createSocket.js

```diff
@@ -2,21 +2,27 @@
 /* global __resourceQuery, __webpack_dev_server_client__ */
 
 const url = require('url');
+const loadWHMEventSource = require('./WHMEventSource');
 
 // This adds support for custom WDS socket transportModes
 // In the future, we should add support for custom clients to better support WDM
 let SocketClient;
 if (typeof __webpack_dev_server_client__ !== 'undefined') {
   SocketClient = __webpack_dev_server_client__;
-} else {
-  SocketClient = require('./SockJSClient');
+  // } else {
+  //   SocketClient = require('./SockJSClient');
 }
```

Comment:
So the main issue that webpack-hot-middleware experiences, is that since `__webpack_dev_server_client__` does not get injected, we trigger this `else` statement. Which also ends up crashing by failing to connect. After the third time the Socket tries to reconnect, it returns html, which then it's internals try to call `res.json()` and crash, prompting the Error screen.

So i'm not really sure what the solution is here. Nor do I really understand what the use case for this custom SocketClient is. Do people really use this? I wonder if they do, if we could require a configuration option to be passed in for this. 

Would love your thoughts here, as this is the main blocker for adoption in webpack-hot-middleware

Reply:
The custom SockJS client is for backwards compatibility with WDS versions before 3.8.0. Before that, WDS will return a raw SockJS client (`onClose` is expressed as `onclose`, and `onMessage` are expressed as `onmessage`, thus breaking code below).

I guess we could fix the compatibility here by detecting the methods listed above and add the wrapper only when needed as a workaround.

Reply:
@pmmmwh good to know. I didn't know that. So let me explain why this wasn't working with webpack-hot-middleware with your current implementation.

[In this if/else statement](https://github.com/pmmmwh/react-refresh-webpack-plugin/blob/master/src/runtime/createSocket.js#L9-L13) the else statement is **always** hit when using webpack-hot-middleware because it does not inject that global variable `__webpack_dev_server_client__`.

When we hit that else statement, the sockjs socket is initialized and starts trying to create a connection to some endpoint. This endpoint is never valid in a webpack-hot-middleware setup. So what happens then is that it tries to connect to that endpoint a couple times (has a retry mechanism internally). Then when it finally decides it can't connect, it emits an error which triggers the ErrorOverlay. 🚫 

So currently any use of this plugin in webpack-hot-middleware always results in the ErrorOverlay being displayed for a SockJS connection error.

Thus the fix is to only initialized the `__webpack_dev_server_client__` if we are in a webpack-dev-server mode. This is easy when the global variable is around because we can immediately detect it and know that we should use it to listen for compile messages.

When we are in webpack-hot-middleware, we do not have this sort of mechanism. The global variable it does install is not installed immediately and makes it hard to react to. The best way to do this is just to import the package directly and use it. This would be a noop in other environments because the client wouldn't do anything. But still we guard against it.

Lastly, this leave the legacy webpack setup. I was hoping there was some heuristics at build time to know if that setup was involved, but I couldn't find anything. The best option I see is that we had a configration flag for using legacy websockets. This is the implementation I have currently taken and would like your feedback and discussion on. 

The new configuration is `new ReactRefreshPlugin({ useLegacyWebsockets: true })`. What do you think about this approach?

Reply:
I think a reasonably heuristic to test for the legacy socket is via `client.onmessage` - in WDS < 3.8 it will be defined, and in WDS >= 3.8 it will be undefined (`client.onMessage` will be defined, instead).

There are actually two problems here (sorry for the confusion 😭):
1. Support for legacy socket implementation in WDS (which can be detected via the heuristic above)
2. Support for non-inline mode WDS usage - as seen in this [line](https://github.com/webpack/webpack-dev-server/blob/2da4a1e0328f8271a0b9b5ec40dcafeb231cfb09/client-src/default/socket.js#L8), WDS do not inject the client (because the whole runtime is injected via an `iframe`). While this is [non-suggested usage](https://webpack.js.org/configuration/dev-server/#devserverinline), I think your implementation actually solved this issue.

Nonetheless, I think both are out of scope for this PR, but since you've already added the flag, maybe rename that to `useLegacyWDSSockets` to indicate that it is exclusive to WDS? Then we can get this PR merged, and I can take on the first issue in another PR 🎉

Reply:
Awesome! I'll add this right now and add it to the documentation! Give me ~30m!

Reply:
Done! Ready for you to review/merge. Thanks @pmmmwh !